src/Uno.UI/Helpers/WeakEvents/WeakEventManager.cs

```diff
@@ -6,130 +6,110 @@
 
-		public void HandleEvent(string eventName)
+		var target = handler.Target;
+		if (target is null)
 		{
-			if (_eventHandlers.TryGetValue(eventName, out List<Subscription>? target))
-			{
-				for (int i = 0; i < target.Count; i++)
-				{
-					Subscription subscription = target[i];
-					bool isStatic = subscription.Subscriber == null;
-					if (isStatic)
-					{
-						// For a static method, we'll just pass null as the first parameter of MethodInfo.Invoke
-						subscription.Handler.Invoke(null, null);
-						continue;
-					}
-
-					object? subscriber = subscription.Subscriber?.Target;
-
-					if (subscriber == null)
-					{
-						// The subscriber was collected, so there's no need to keep this subscription around
-						target.Remove(subscription);
-					}
-					else
-					{
-						subscription.Handler.Invoke(subscriber, null);
-					}
-				}
-			}
+			AddTo(_staticHandlers ??= new(), eventName, handler);
 		}
-
-		public void RemoveEventHandler(Action? handler, [CallerMemberName] string eventName = "")
+		else
 		{
-			if (string.IsNullOrEmpty(eventName))
-				throw new ArgumentNullException(nameof(eventName));
+			AddTo(_handlersPerTarget.GetValue(target, static _ => new()), eventName, handler);
+		}
+	}
 
-			if (handler == null)
-				throw new ArgumentNullException(nameof(handler));
+	public void RemoveEventHandler(Action handler, [CallerMemberName] string eventName = "")
+	{
+		if (string.IsNullOrEmpty(eventName))
+			throw new ArgumentNullException(nameof(eventName));
 
-			RemoveEventHandler(eventName, handler.Target, handler.GetMethodInfo());
-		}
+		if (handler == null)
+			throw new ArgumentNullException(nameof(handler));
 
-		private void AddEventHandler(string eventName, object? handlerTarget, MethodInfo methodInfo)
+		var target = handler.Target;
+		if (target is null)
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? targets))
-			{
-				targets = new List<Subscription>();
-				_eventHandlers.Add(eventName, targets);
-			}
-			else
-			{
-				targets.RemoveAll(subscription => subscription.Subscriber is { IsAlive: false });
-			}
-
-			if (handlerTarget == null)
+			if (_staticHandlers is not null)
 			{
-				// This event handler is a static method
-				targets.Add(new Subscription(null, methodInfo));
-				return;
+				RemoveFrom(_staticHandlers, eventName, handler);
 			}
-
-			targets.Add(new Subscription(new WeakReference(handlerTarget), methodInfo));
 		}
-
-		private void RemoveEventHandler(string eventName, object? handlerTarget, MethodInfo methodInfo)
+		else if (_handlersPerTarget.TryGetValue(target, out var handlersPerEvent))
 		{
-			if (!_eventHandlers.TryGetValue(eventName, out List<Subscription>? subscriptions))
-				return;
+			RemoveFrom(handlersPerEvent, eventName, handler);
+		}
+	}
 
-			for (int n = subscriptions.Count - 1; n >= 0; n--)
+	public void HandleEvent(string eventName)
+	{
+		_enumeratingHandlersOfEvent = eventName; // Note: We do not support re-entrancy!
```

Comment:
I think re-entrancy should actually be allowed. It may not have been allowed in the existing implementation, but that means the existing implementation is not correct :)

Reply:
By re-entrancy here I mean that we are not allowed to raise an event while we are already raising an event ... but actually re-thinking about it, it could be easily fixed without perf impact, will fix it ^^